qiskit/circuit/parametertable.py

```diff
@@ -12,43 +12,140 @@
 """
 Look-up table for variable parameters in QuantumCircuit.
 """
-import warnings
 import functools
-from collections.abc import MutableMapping, MappingView
+import warnings
+from collections.abc import MappingView, MutableMapping, Set
+from typing import MutableSequence
+
+
+class ParameterReferences(MutableSequence, Set):
+    """An (insertion) ordered set of instruction parameter slot references.
+    Items are expected in the form ``(instruction, param_index)``. Membership
+    testing is overriden such that items that are otherwise value-wise equal
+    are still considered distinct if their ``instruction``\\ s are referentially
+    distinct.
+
+    Items are ordered by insertion, and may be read or deleted by sequence index.
+    However, update by index and insert by index are not supported, since
+    these actions inherently violate insertion order.
+    """
+
+    def _instance_key(self, ref):
+        return (id(ref[0]), ref[1])
+
+    def __init__(self, *refs) -> None:
+        self._data = []
+        self._instance_ids = set()
+
+        for ref in refs:
+            if not isinstance(ref, tuple) or len(ref) != 2:
+                raise ValueError("refs must be in form (instruction, param_index)")
+            k = self._instance_key(ref)
+            if k not in self._instance_ids:
+                self._data.append(ref)
+                self._instance_ids.add(k)
+
+    @classmethod
+    def from_iterable(cls, iterable):
+        """Create a new instance from an iterable of ``(instruction, param_index)``
+        tuples."""
+        return cls(*iterable)
```

Comment:
This is entirely minor, but I'm not really sure why `__init__` takes a variadic argument and this "from iterable" form isn't just the only (and default) constructor.

Reply:
IIRC, I did this to support syntax like `ParameterReferences(ref1, ref2)`, which is used in a couple of places. I believe there may be another subtlety that I'm forgetting in the choice of adding a named constructor like this, having to do with not being able to write a single `__init__` in a way that can disambiguate iterables from references.

I remember taking inspiration from [itertools.chain](https://docs.python.org/3/library/itertools.html#itertools.chain).

Let me know if you'd like me to change this.

Reply:
I think `itertools.chain` is perhaps solving a slightly different problem - in that, it's common to want to construct the object with "hardcoded" arguments, like
```python
for key in itertools.chain(left_keys, right_keys):
    ...
```
where it's convenient syntactically to not require the extra `[]` or `()` in the constructor, and it's generic over its inputs, so it's valid to chain together iterables of iterables (i.e. they can't infer from the input, but you could if you wanted to).  `chain` also needs to support infinite/lazy iterable inputs, which is where `from_iterable` really becomes necessary.

In this, the only situation I can think of where you'd create one with a "hardcoded" input is the singleton case, but for all the rest you want to iterate through an unknown number of references to build it up, so I'd perhaps optimise its syntax around that.